### ISSUE TITLE
TreeGrid - removal of child is not correctly reflected in view

### ISSUE THREAD

**Issue by pellmann** [vaadin/vaadin-grid#1819]
When a node is removed, it is removed also in the view. But with the last one this does not work. The parent looses the marker for children but the last child is viewed. 

This is reproducible with the code from https://github.com/vaadin/vaadin-grid-flow/issues/1087

Nodes can be removed with the context menu.

Tested with Vaadin 14.3.3, Java 11.0.6, OSX 10.15.6, Newest Chrome and with Spring Boot 2.2.0

**Comment by jcgueriaud1** [comment]
I tested in Vaadin 14.4.2. This example:

```
@Route("TreeGridView")
public class TreeGridView extends VerticalLayout {

    public TreeGridView() {
        TreeGrid<String> grid = new TreeGrid<>();
        grid.addHierarchyColumn(String::toString).setHeader("Name");

        String root = "Root";
        String child = "child";
        String subChild = "Sub-child";
        String child2 = "child2";
        String subChild2 = "Sub-child2";
        String child3 = "child3";

        TreeData<String> td = grid.getTreeData();
        td.addItem(null, root);
        td.addItem(root, child);
        td.addItem(child, subChild);
        td.addItem(root, child2);
        td.addItem(child2, subChild2);
        td.addItem(root, child3);
        grid.expand(root, child, subChild, child2);

        Button button = new Button("Remove first child", e -> {
            td.removeItem(child);
            grid.getDataProvider().refreshItem(root);
        });

        Button button2 = new Button("Remove 2nd child", e -> {
            td.removeItem(child2);
            grid.getDataProvider().refreshItem(root);
        });

        Button buttonR = new Button("Remove recur first child", e -> {
            td.removeItem(child);
            grid.getDataProvider().refreshItem(root, true);
        });

        Button buttonR2 = new Button("Remove recur 2nd child", e -> {
            td.removeItem(child2);
            grid.getDataProvider().refreshItem(root, true);
        });
        add(grid, button, button2, buttonR, buttonR2);
    }
}
```

If you click on "Remove recur first child" then the treegrid is showing the wrong data (see the attached screenshot)
).

The child3 contains sub-child2, child2 contains sub-child.

![treegrid](https://user-images.githubusercontent.com/51313578/101137019-d8d5c200-3616-11eb-8a5e-296e310c3a8a.png

**Comment by jcgueriaud1** [comment]
I've tried to collapse/expand, but it doesn't help:
```
Button buttonR = new Button("Remove recur first child", e -> {
    grid.collapse(child);
    td.removeItem(child);
    grid.getDataProvider().refreshItem(root, true);
    grid.collapse(root);
    grid.expand(root);
});
```

**Comment by tulioag** [comment]
Updated the title because it fails in more cases.

**Comment by tomivirkki** [comment]
This seems to work
```
Button buttonR = new Button("Remove recur first child", e -> {
  td.removeItem(child);
  grid.getDataProvider().refreshItem(root, true);
  grid.getElement().executeJs("this.clearCache()");
});
```
...but the issue should still be fixed

**Comment by tarekoraby** [comment]
@tulioag unfortunately, at least some of the issues described in this ticket are still not fixed in 14.4.9. Removing the last child is still not working:
```
TreeGrid<String> grid = new TreeGrid<>();
grid.addHierarchyColumn(String::toString).setHeader("Name");

String parent = "parent";
String childOne = "child one";
String childTwo = "child two";

TreeData<String> td = grid.getTreeData();
td.addItem(null, parent);
td.addItem(parent, childOne);
td.addItem(parent, childTwo);

grid.expand(parent);

Button removeChildOne = new Button("Remove Child One", e -> {
    td.removeItem(childOne);
    grid.getDataProvider().refreshItem(parent, true);
});

Button removeChildTwo = new Button("Remove Child Two", e -> {
    td.removeItem(childTwo);
    grid.getDataProvider().refreshItem(parent, true);
});
```


The workaround mentioned by @tomivirkki is still valid though
